draft-irtf-cfrg-vdaf.md

```diff
@@ -1363,6 +1376,57 @@ class PrgSha3(Prg):
 ~~~
 {: title="Definition of PRG PrgSha3."}
 
+### PrgFixedKeyAes128 {#prg-fixed-key-aes128}
+
+While PrgSha3 as described above can be securely used in all cases where a Prg
+is needed in the VDAFs described in this document, there are some cases where
+a more efficient instantiation based on fixed-key AES is possible. For now, this
+is limited to the Prg used inside the Idpf {{idpf}} implementation in Poplar1
+{{idpf-poplar}}. It is NOT RECOMMENDED to use this Prg anywhere else.
+See Security Considerations {{security}} for a more detailed discussion.
+
+~~~
+class PrgFixedKeyAes128(Prg):
+    # Associated parameters
+    SEED_SIZE = 16
+
+    def __init__(self, seed, custom, binder):
+        self.length_consumed = 0
+
+        # Use SHA-3 to derive a key from the binder and customization
+        # strings. Note that the AES key does not need to be kept
+        # secret from any party. However, when used with IpdfPoplar,
+        # we require the binder to be a random nonce.
+        #
+        # Implementation note: This step can be cached across PRG
+        # evaluations with many different seeds.
+        self.fixed_key = cSHAKE128(binder, 16, b'', custom)
+        self.seed = seed
+
+    def next(self, length: Unsigned) -> Bytes:
+        offset = self.length_consumed % 16
+        new_length = self.length_consumed + length
+        block_range = range(
+            int(self.length_consumed / 16),
+            int(new_length / 16) + 1)
+        self.length_consumed = new_length
+
+        hashed_blocks = [
+            self.hash_block(xor(self.seed, to_le_bytes(i, 16))) \
+                         for i in block_range
+        ]
+        return concat(hashed_blocks)[offset:offset+length]
+
+    # The multi-instance tweakable circular correlation-robust hash function of
+    # [GKWWY20] (Section 4.2).
+    #
+    # Function `AES128(key, block)` is the AES-128 blockcipher.
+    def hash_block(self, block):
+        lo, hi = block[:8], block[8:]
+        sigma = hi + xor(hi, lo)
```

Comment:
Is the '+' operator here a concatenation?  It might be better to make that explicit because I came away with the distinct impression that you were losing the upper bits of the value here.

Reply:
sigma is a function, not a variable, so perhaps you could frame it that way

Reply:
* Yes, "+" means concatenation here. We can replace with "concat()" for clarity?
* Actually, sigma is meant to be a variable (it is input to AES, then xored with the output).

Reply:
https://eprint.iacr.org/2019/074.pdf Section 7.3 (last paragraph, "Instantiating $sigma$") has it as a function.  That function needs to be a linear orthomorphism, as this is.  The value that is input to AES is the output of that function (which I guess you can call sigma too if you like).

Reply:
Agreed, explicitly calling `concat` would be better here. We can call the variable `sigma_block` or so to emphasize that it is the function sigma evaluated on `block`?